contracts/distribution.js

```diff
@@ -0,0 +1,301 @@
+/* eslint-disable no-await-in-loop */
+/* eslint no-underscore-dangle: ["error", { "allow": ["_id"] }] */
+/* global actions, api */
+
+// limit recipients per distribution batch
+const MAX_RECIPIENTS = 40;
+
+actions.createSSC = async () => {
+  const tableExists = await api.db.tableExists('batches');
+  if (tableExists === false) {
+    await api.db.createTable('batches');
+    await api.db.createTable('params');
+
+    const params = {};
+    params.distCreationFee = '500';
+    params.distUpdateFee = '250';
+    await api.db.insert('params', params);
+  }
+};
+
+actions.updateParams = async (payload) => {
+  if (api.sender !== api.owner) return;
+
+  const {
+    distCreationFee,
+    distUpdateFee,
+  } = payload;
+
+  const params = await api.db.findOne('params', {});
+
+  if (distCreationFee) {
+    if (!api.assert(typeof distCreationFee === 'string' && !api.BigNumber(distCreationFee).isNaN() && api.BigNumber(distCreationFee).gte(0), 'invalid distCreationFee')) return;
+    params.distCreationFee = distCreationFee;
+  }
+  if (distUpdateFee) {
+    if (!api.assert(typeof distUpdateFee === 'string' && !api.BigNumber(distUpdateFee).isNaN() && api.BigNumber(distUpdateFee).gte(0), 'invalid distUpdateFee')) return;
+    params.distUpdateFee = distUpdateFee;
+  }
+
+  await api.db.update('params', params);
+};
+
+async function processBatch(batch, symbol, isFlush = false) {
+  let balance;
+  let balanceInd;
+  let payout;
+
+  for (let i = 0; i < batch.tokenBalances.length; i += 1) {
+    const curBalance = batch.tokenBalances[i];
+    if (curBalance.symbol === symbol) {
+      balanceInd = i;
+      balance = curBalance.quantity;
+      break;
+    }
+  }
+
+  for (let i = 0; i < batch.tokenMinPayout.length; i += 1) {
+    const minPayout = batch.tokenMinPayout[i];
+    if (minPayout.symbol === symbol) {
+      payout = minPayout.quantity;
+      break;
+    }
+  }
+
+  if (balance !== undefined && payout !== undefined
+    && (api.BigNumber(balance).gt(api.BigNumber(payout)) || isFlush === true)) {
+    // pay out token balance to recipients by configured share percentage
+    for (let i = 0; i < batch.tokenRecipients.length; i += 1) {
+      const recipient = batch.tokenRecipients[i];
+      const recipientShare = api.BigNumber(balance).multipliedBy(recipient.pct / 100).toFixed(3);
+      if (await api.transferTokens(recipient.account, symbol, recipientShare, recipient.type)) {
```

Comment:
You probably want to check the return value of transferTokens for success.

And we will want to determine what the protocol is on any failures. Should it freeze its state, or should it just fire and forget, and rely on external monitoring to determine if this contract is not behaving properly. One thing that is difficult here is that you dont' have a mechanism of an all-or-nothing rollback mechanism for DB. @bt-cryptomancer , what do you think?

Reply:
It's very unlikely that a transfer will fail, but it's still good practice to do a sanity check on the result to verify success (see other contracts for examples of how to do this).  In the event of a failure, I would record the error with an api.assert but otherwise allow the processing to continue, rather than stop in some halfway complete state.

Reply:
If we encounter a failed transfer, we can keep the quantity in the batch's balance and thereby reattempt distributing it next time it runs. I have to update my test which always expects the batch to get fully emptied on a payment run, but that seems like the most preferable option. Recipients could be updated later and this leftover balance would eventually distribute.

Reply:
Okay that sounds fine.